File: typescript-kata/src/trees/binary-tree/bfs.ts
import { Queue } from "src/data-structures/queue";
import { BinaryNode, BinaryTree } from "./node";

export function BreadthFirstSearch(
  tree: BinaryTree<number>,
  needle: number,
): boolean {
  if (!tree.root) {
    return false;
  }
  let queue = new Queue(); // In the interview we can use [] as queue
  queue.enqueue(tree.root);
  while (queue.length) {
    const node = queue.deque() as BinaryNode<number>;
    if (node.val === needle) {
      return true;
    }
    if (node.right) {
      queue.enqueue(node.right);
    }

    if (node.left) {
      queue.enqueue(node.left);
    }
  }

  return false;
}

File: go-kata/internal/trees/binary_tree/bfs.go
package binary_tree

func BreadthFirstSearch(root BinaryNode[int], needle int) bool {
	q := []BinaryNode[int]{}

	q = append(q, root)
	for len(q) > 0 {
		var node BinaryNode[int]
		node, q = q[0], q[1:]

		if node.Val == needle {
			return true
		}

		if node.Left != nil {
			q = append(q, *node.Left)
		}

		if node.Right != nil {
			q = append(q, *node.Right)
		}

	}

	return false
}
